### ISSUE TITLE
Policy Gradient

### ISSUE THREAD

**Issue by MXD6** [aai-institute/tianshou#777]
Should parameter updates be outside the loop?

<img width="686" alt="image" src="https://user-images.githubusercontent.com/31929862/203955554-c8051f3f-7fb6-44bd-8a33-1ad837768145.png">

**Comment by MXD6** [comment]
Otherwise, when `result = self(minibatch)` is called, the network has already been updated.

**Comment by MXD6** [comment]
The implementation of NPG seems to have the same problem.

<img width="666" alt="image" src="https://user-images.githubusercontent.com/31929862/204082222-2104d258-8412-4261-91be-59d5d14d4811.png">

**Comment by janofsun** [comment]
`result = self(minibatch` will not update the param of the network. `optim.step()` actually update them.

> Otherwise, when `result = self(minibatch)` is called, the network has already been updated.

**Comment by Trinkle23897** [comment]
`self(minibatch)` calls `self.__call__`, which will be redirect to `self.forward` function if it's a pytorch `nn.Module`. In this case `self(...)` won't call `self.learn` and only does a forward-only operation.

**Comment by MXD6** [comment]
In the first loop, the network parameters are updated through `self.optim.step`. In the second loop, the `dist` obtained through `self (minibatch)` is not a sampling dist.

**Comment by MXD6** [comment]
@Trinkle23897

**Comment by Trinkle23897** [comment]
@ChenDRAG

**Comment by ChenDRAG** [comment]
> Otherwise, when `result = self(minibatch)` is called, the network has already been updated.

Yes, the network may have been updated when result = `self(minibatch)` is called. Is there any problem with this?

I'm not sure, but my guess about your concern is that "We should only update the network once, for each data collection". However, this is difficult when you collect very much data each time (without a minibatch there could be OoM problem)

The approximation is that the network parameters won't shift far away for a few updates, and we empirically find that ``more update, smaller batch size'' could lead to better performance (in PPO, for instance). @MXD6

**Comment by MXD6** [comment]
Thank you for your answer! @ChenDRAG

**Comment by Trinkle23897** [comment]
Oh forgot to reply, I think you can always use repeat=1 and batch_size=len(batch) to avoid gradient estimation inconsistency
